### PR TITLE
support for operation on installer images rather than ISOs

### PR DESCRIPTION
**Pull request by mwhudson**
This still operates fairly similarly to ISOs in that it makes a new image, even though for an image working in place would be possible. cc @dbungert @sil2100 @xypron

### CODE REVIEW

**Comment by mwhudson** [comment]
Running this on my amd64 laptop trying to update the subiquity snap in a riscv64 installer image still fails, but not for any reason that's related to images vs ISOs as far as I can tell:

```
(image-support)mwhudson@anduril:~/src/livefs-editor$ sudo PYTHONPATH=~/src/livefs-editor python3 -m livefs_edit ~/images/riscv64.img ~/images/riscv64-new.img --setup-rootfs --inject-snap ~/images/subiquity_22.02.2+git97.f9ce5370_riscv64.snap 
set up loop device /dev/loop100 backing /home/mwhudson/images/riscv64.img
found live ext4 filesystem on /dev/loop100p1
running setup-rootfs with arguments {}
  running unpack-initrd with arguments {}
running inject-snap with arguments {'snap': '/home/mwhudson/images/subiquity_22.02.2+git97.f9ce5370_riscv64.snap'}
running repack hooks
  running snap-preseed...
error: cannot mount /var/lib/snapd/seed/snaps/snapd_13398.snap at /tmp/snapd-preseed in preseed mode: exit status 32
'mount -t squashfs -o ro,x-gdu.hide,x-gvfs-hide /var/lib/snapd/seed/snaps/snapd_13398.snap /tmp/snapd-preseed' failed with: mount: /tmp/snapd-preseed: failed to setup loop device for /var/lib/snapd/seed/snaps/snapd_13398.snap.

Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/mwhudson/src/livefs-editor/livefs_edit/__main__.py", line 72, in <module>
    main(sys.argv[1:])
  File "/home/mwhudson/src/livefs-editor/livefs_edit/__main__.py", line 64, in main
    ctxt.repack(destpath)
  File "/home/mwhudson/src/livefs-editor/livefs_edit/context.py", line 242, in repack
    hook()
  File "/home/mwhudson/src/livefs-editor/livefs_edit/actions.py", line 244, in _preseed
    run(['/usr/lib/snapd/snap-preseed', rootfs])
  File "/home/mwhudson/src/livefs-editor/livefs_edit/__init__.py", line 7, in run
    return subprocess.run(cmd, check=check, **kw)
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/lib/snapd/snap-preseed', '/tmp/tmpp9z_oh7h/rootfs']' returned non-zero exit status 1.
```

(I'm not sure snap-preseed is expected to work cross-arch like this?). I guess for testing for now we can leave the preseed call out (we still need the --reset bit but that seems to work).

File: livefs_edit/context.py
Comment (sil2100) [inline livefs_edit/context.py:220]: Ok, I like this: simple but 'just works'. When preparing my own branch I was thinking that doing guesswork on where the live-media fs is would be a bit too involving, but basically just looking for a partition with casper squashfses indeed is enough!

**Review by sil2100 — APPROVED** [review]
This looks good to me, thank you Michael! For the riscv64 experiments I did, I also prepared some 'ugly modifications' in https://github.com/sil2100/livefs-editor/tree/add-non-iso-support , but this basically required people to mount the live-media rootfs partitions themselves - not to mention some ugliness by doing the final copying in teardown. So I'm much happier with the clean approach you designed.

And yes, `snap-preseed` is not cross-portable. All the testing for my branch I was doing was basically on an amd64 hand-crafted .img file (with Heinrich giving it a spin on riscv64 hardware later on). So I'd recommend waiting for Heinrich's or Alexandre's final feedback before merging this in.

File: livefs_edit/context.py
Comment (mwhudson) [reply livefs_edit/context.py:220]: It's what casper does! (Well, more or less -- casper supports some non-squashfs files too but blah).

**Comment by mwhudson** [comment]
> And yes, `snap-preseed` is not cross-portable.

Maybe we can detect this and not preseed in this case then. (In a different branch!)

**Comment by xypron** [comment]
The problem seems to be related to the file names
```
sudo LIVEFS_EDITOR=~/livefs-editor/ ./scripts/inject-subiquity-snap.sh ~/in.img ../my.snap ~/out.img
```
does not show the problems that I see with
```
sudo LIVEFS_EDITOR=~/livefs-editor/ ./scripts/inject-subiquity-snap.sh ~/installer-base.img ../subiquity_22.02.2+git97.f9ce5370_riscv64.snap ~/installer-target.img
```

**Comment by mwhudson** [comment]
```
[18652.750896] /home/ubuntu/subiquity_22.02.2+git97.f9ce5370_riscv64.snap: Can't open blockdev
```

This is a message from the kernel I think? I don't know what's going on but it seems a bit unlikely to be file name related somehow. Is it reproducible?

**Comment by mwhudson** [comment]
This seems to be working well enough so I'll merge this.

**Pull request merged**
